File: apps/desktop/src/api/categories.ts
import { invoke } from "@tauri-apps/api/core";
import type { Category, CategoryInput } from "./types";

/**
 * Typed wrappers around the `commands::categories` Tauri commands. This is the only module that
 * should call `invoke()` for categories — components call these functions instead.
 */

export function listCategories(includeInactive: boolean): Promise<Category[]> {
  return invoke("list_categories", { includeInactive });
}

export function createCategory(input: CategoryInput): Promise<Category> {
  return invoke("create_category", { input });
}

export function updateCategory(id: number, input: CategoryInput): Promise<Category> {
  return invoke("update_category", { id, input });
}

export function archiveCategory(id: number): Promise<void> {
  return invoke("archive_category", { id });
}

export function reactivateCategory(id: number): Promise<void> {
  return invoke("reactivate_category", { id });
}

export function deleteCategory(id: number): Promise<void> {
  return invoke("delete_category", { id });
}

File: apps/desktop/src-tauri/src/commands/categories.rs
use tauri::State;

use crate::db::repositories::categories::{self, Category, CategoryInput};
use crate::error::{AppError, AppResult};
use crate::DbState;

#[tauri::command]
pub fn list_categories(db: State<DbState>, include_inactive: bool) -> AppResult<Vec<Category>> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    categories::list(&conn, include_inactive)
}

#[tauri::command]
pub fn create_category(db: State<DbState>, input: CategoryInput) -> AppResult<Category> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    categories::create(&conn, input)
}

#[tauri::command]
pub fn update_category(db: State<DbState>, id: i64, input: CategoryInput) -> AppResult<Category> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    categories::update(&conn, id, input)
}

#[tauri::command]
pub fn archive_category(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    categories::set_active(&conn, id, false)
}

#[tauri::command]
pub fn reactivate_category(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    categories::set_active(&conn, id, true)
}

#[tauri::command]
pub fn delete_category(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    categories::delete(&conn, id)
}
